pyxem/hyperspy_extension.yaml

```diff
@@ -83,6 +83,12 @@ signals:
     dtype: real
     lazy: False
     module: pyxem.signals.diffraction_vectors
+  LazyDiffractionVectors:
+    signal_type: diffraction_vectors
```

Comment:
My opinion is that vectors are a special type of Signal with:

- 0 Signal Dimensions - A ragged array with a 2d array of variable number of rows and fixed columns at every point.
- 1 Signal Dimensions- An array of columns with 1 vector for each position 
- 2 Signal Dimensions- An array of columns with multiple vectors for each position. Fixed number of rows and columns  

Anything beyond this shouldn't be supported as it starts to become a little unphysical.  In all cases each signal has a constant number of columns with the variable being the number of rows.

You're right that the signal_type should probably be "diffraction" in the longer term as that would help with things like returning a vectors object from something like the `find_peaks` function. A `vector` key in the signal specification would go a long way there. 

For a `hyperspy` specific solution, we probably need to:

1. Clean up the axes specification and add the ability to add Vector axes if the signal is a `vector`
2. Define `icol` and `irow` as slicers which work with the 0-, 1- and 2-D  cases. (This should mostly work right now but should be a little more bullet proof)
3. Define a BaseVectorSignal similar to the BaseSignal as well as a VectorSignal1D and a VectorSignal2D. 

The biggest sticking point is figuring out how to handle the axes etc.

The `icol` and `irow` could probably be removed in favor of some changes to `isig`.  For example:

`s.isig[:,s.isig["x"]>0.5]` would work as well as `s.irow[s.icol["x"]>0.5]`.  That is kind of what I was trying to do in https://github.com/hyperspy/hyperspy/pull/3055 but I don't know if that implementation was super clean. I guess that would also make it truly n-dimensional so it's probably the proper implementation.

Reply:
Not sure what the proper thing to do is.  We can always hold some of these changes for a 2.1.0 release but I also think they are some nice features and it would be nice to add them sooner than later.  I can clean up and make another pass through  https://github.com/hyperspy/hyperspy/pull/3031  another time which adds minimal support for vector axes.  That is pretty much good to go except we have to figure out if we should/shouldn't support non-increasing/decreasing data.